### ISSUE TITLE
How to update the environment during the execution of a command?

### ISSUE THREAD

**Issue by jakubkaczor** [nushell/nushell#14364]
### Question

Does `def --env` merge the environments at the end of execution? Shouldn't it work directly on the existing environment, outer scope?

### Additional context and details

I try to implement the following POSIX shell code
```sh
pushd $(mktemp --directory)
$EDITOR file
popd
```
using the following Nushell command
```nu
def --env temp-edit [] {
  use std/dirs
  dirs add (mktemp --directory)
  ^$env.EDITOR file
  dirs drop
}
```
The behavior is not the same, what can be noticed when using terminal multiplexer Zellij. In Zellij, after running `temp-edit` the editor opens up as expected, having working directory as the temporary directory, but the Zellij's working directory is the same as before running `temp-edit` --- contrary to what happens with POSIX code. I suspect this is because Nu updates the scope where `temp-edit` was invoked only after `temp-edit` terminates. If you removed `dirs drop` from the last command, after exiting the editor, you would end up in the temporary directory also in Zellij, but not during `^$env.EDITOR file`.

Is this expected behavior? Why? If so, is there some other recommended way of achieving the behavior I expected?

**Comment by fdncred** [comment]
I haven't tried this but it seems like if you want all the cding to happen within the custom command and not change directory when it's done, you don't need the `--env`. I think without the --env you'll be in the folder you started but the cd stuff happens only within the custom command.

Maybe I've misunderstood.

**Comment by jakubkaczor** [comment]
> Maybe I've misunderstood.

Hmm, perhaps.

> I haven't tried this but it seems like if you want all the cding to happen within the custom command and not change directory when it's done, you don't need the `--env`. I think without the --env you'll be in the folder you started but the cd stuff happens only within the custom command.

That is right. I already tried just `cd`ing inside a command without `--env`, and the result before and after the command were the same, but the problem with Zellij during the execution of the command remains.

I am not sure how Zellij operates, but I suspect it may read the `PWD` environment variable from the scope it was started in, so the variable in the same scope would need to be changed. Correct me if I am wrong, but unless `temp-edit` exits, outer scope is not updated with the new value for `PWD` when using `--env`.

**Comment by IanManske** [comment]
Yes, `def --env` merges environment changes from a command once the command finishes execution.

However, I'm not seeing a difference in behavior between the POSIX and Nushell code inside Zellij. What version of Nushell on you on?

**Comment by jakubkaczor** [comment]
`nu --version` returns `0.99.1`. After reading your message I tried
```sh
temp-edit() {
  pushd $(mktemp --directory)
  $EDITOR test.txt
  popd
}
```
then invoked `temp-edit`, and I also couldn't reproduce the mismatch. This was because I tried this inside `zsh` running inside `nu` running `zellij`. After I set `SHELL=/usr/bin/(zsh|bash)` and switched to a new TTY, I could reproduce the mismatch with both, zsh and bash. How do you run POSIX shell and Zellij? Perhaps similarly to what I described at first, do you?

**Comment by IanManske** [comment]
This is what I see:
```nu
# Started a new terminal with SHELL=/usr/bin/nu, then ran `zellij`
> pwd
# prints home directory, zellij shows home directory

> def --env temp-edit [] {
∙   use std/dirs
∙   dirs add (mktemp --directory)
∙   ^pwd
∙   dirs drop
∙ }

> temp-edit
/tmp/tmp.CrGGl3wph2

> pwd
# prints home directory, zellij shows home directory
```

```bash
# Started a new terminal with SHELL=/usr/bin/bash, then ran `zellij`
$ pwd
# prints home directory, zellij shows home directory

$ temp-edit() {   pushd $(mktemp --directory);   pwd;   popd; }

$ temp-edit
/tmp/tmp.iYW9bIqeMI ~
/tmp/tmp.iYW9bIqeMI
~

$ pwd
# prints home directory, zellij shows home directory
```

In what circumstance(s) are you seeing the mismatch?

**Comment by jakubkaczor** [comment]
If I interpret correctly, you inspect with the `pwd` _after_ the `temp-edit` finishes, which gives the same results for me. I notice the difference in the following. In a POSIX shell, during the execution of `temp-edit` the path displayed in Zellij's pane's top-left corner changes, and new floating panes started with `Ctrl + p, w` start a shell with a different path from home. To be precise, with the temporary directory created with the `mktemp --directory` that was pushed with `pushd`, and this is what I expect. This does not happen in the corresponding Nu code. There, Zellij always shows my home directory path, and, in new panes, starts shells at my home directory path.

**Comment by jakubkaczor** [comment]
I get the same results running your code, but this is expected. As I said, the `$EDITOR` displays the temporary directory with the editor's pwd command, which is equivalent to your `pwd`. The difference is with Zellij, which runs in the outer scope of `temp-edit`, that is, the scope where `temp-edit` is invoked, not in the scope of `temp-edit`.

**Comment by IanManske** [comment]
Oh ok, I see the issue now. Nushell has support for parallelism via `par-each` and a few other commands. So, to preserve isolation between threads, updates to the environment are applied to Nushell's emulated environment and not to the environment of the actual shell process. Only once a REPL command/entry fully finishes (shortly before the new prompt is displayed), is the PWD updated for the shell process. This is probably why PWD is not up to date in the middle of execution (zellij most likely uses the PWD of the nushell process).

**Comment by jakubkaczor** [comment]
> This is probably why PWD is not up to date in zellij during the middle of execution (zellij most likely uses the PWD of the nushell process).

That is what I suspected.

> Nushell has support for parallelism via par-each and a few other commands. So, to preserve isolation between threads, updates to the environment are applied to Nushell's emulated environment and not to the environment of the actual shell process.

Right, joining environments upon joining threads makes sense, but actually, it is weird to me that such “impure” commands, such as defined with `--env`, are allowed in such scenarios. There can be a conflict after joining, during consolidation of `$env`. Are such concurrent updates defined behavior? I cannot find such information with `help par-each`, neither with `help def`.

**Comment by jakubkaczor** [comment]
Are there scenarios, where executing `--env` commands in parallel is necessary? I don't see any, but it appears that the use case I presented would make use of updating the environment during the execution of a command. Perhaps it would be more suitable to update during execution, and disallow `--env` commands from the use in commands using parallelism? It could be detected before execution. Should I create a separate feature request?

**Comment by IanManske** [comment]
`def --env` commands only affect the current thread which is why they are allowed to be used anywhere. Also, `par-each` takes closures which cannot be declared with `--env`, so any environment changes inside the closure will be discarded. I.e,. there is no join, so there cannot be a join conflict. The documentation should probably explain this in more detail.

The issue is not only `--env` commands but any command that alters the environment like `cd`, `load-env`, or `$env.var = value`. Disallowing all of these inside `par-each` would hinder its usefulness and would be a major burden to keep track of. E.g., can I use this command inside `par-each`? Does this command call other commands that cannot be used inside `par-each`? Etc...

What could be possible though, is to differentiate between the "main thread" and other threads. When code is first executed, there is only one thread, so it would be safe to directly alter the process environment. However, `def` commands without `--env` and any closures would need to start a new scope, altering only the emulated environment. I think this would work, but it could add some complexity to the code and I'm not sure if it's worth the benefits. Feel free to open a feature request though! This might change in the future. E.g., maybe there's other issues caused by delaying environment updates that have yet to surface.

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered.
